### PR TITLE
Support dot notation in projection

### LINKED ISSUE #2430: Support dot notation in query projections
### What should be done?

We should add support for dot notation in query projections.

See relevant test cases with this issue number.

The tricky part is that we have to create an empty document when the given path could not be reached.

```js
db.projections.insertMany([
 {
    a: { b: { c: { d: 123 }, e: 13 }, f: 14 },
    g: 15
  },
  {
    a: { b: { c: [ 1, 2 ], e: 13 }, f: 14 },
    g: 15
  },
  {
    a: { b: 12 }
  },
  {
    a: { b: { c: 12 } } 
  }
]);


db.projections.find({}, {'_id': 0,'a.b.c.d': 0})
[
  { a: { b: { c: {}, e: 13 }, f: 14 }, g: 15 },
  {
    a: { b: { c: [ 1, 2 ], e: 13 }, f: 14 },
    g: 15
  },
  { a: { b: 12 } },
  { a: { b: { c: 12 } } }
]


 db.projections.find({}, {'_id': 0,'a.b.c.d': 1})
[
  {
    a: { b: { c: { d: 123 } } }
  },
  { a: { b: { c: [] } } },
  { a: {} },
  { a: { b: {} } }
]

db.projections.find({}, {'_id': 0,'a.b.c': 0})
[
  { a: { b: { e: 13 }, f: 14 }, g: 15 },
  { a: { b: { e: 13 }, f: 14 }, g: 15 },
  { a: { b: 12 } },
  { a: { b: {} } }
]


db.projections.find({}, {'_id': 0,'a.b.c': 1})
[
  {
    a: { b: { c: { d: 123 } } }
  },
  {
    a: { b: { c: [ 1, 2 ] } }
  },
  { a: {} },
  { a: { b: { c: 12 } } }
]

```

### Where?

internal/handlers/common/projection.go

### Definition of Done

- all handlers updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated;
- something else?

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #2430.

## Readiness checklist

* [ ] I added/updated unit tests.
* [x] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [ ] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/query_projection_compat_test.go [hunk 1]
```diff
@@ -141,33 +145,55 @@ func TestQueryProjectionCompat(t *testing.T) {
...
+			filter:     bson.D{},
+			projection: bson.D{{"v.a.b.c", false}},
+		},
+		"DotNotationArray": {
+			filter:     bson.D{},
+			projection: bson.D{{"v.array.0", true}},
+		},
+		"DotNotationArrayPath": {
+			filter:     bson.D{},
+			projection: bson.D{{"v.0.foo", true}},
```
Comment (chilagrow) [inline integration/query_projection_compat_test.go:196]: This test ensures that we don't used `doc.SetByPath(...)`. In projection array index path does not find anything.

**Review by noisersup — APPROVED** [review]
:+1:

File: internal/handlers/common/projection.go [hunk 2]
```diff
@@ -221,28 +221,270 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
+		return arr, nil
+	default:
+		return nil, pathNotExists
+	}
+}
+
+// isEmpty returns true if projected document at key is empty. It's empty if:
+// - key does not exist in projected document,
+// - projected contains empty document at key, or
+// - projected contains an array which contain empty documents at key.
```
Comment (w84thesun) [inline internal/handlers/common/projection.go:357]:
```suggestion
//   - key does not exist in projected document,
//   - projected contains empty document at key, or
//   - projected contains an array which contain empty documents at key.
```

Comment (w84thesun) [inline internal/handlers/common/projection.go:357]: To make it render as list
```suggestion
//   - key does not exist in projected document,
//   - projected contains empty document at key, or
//   - projected contains an array which contain empty documents at key.
```

**Review by w84thesun — COMMENTED** [review]
Great job implementing all this logic! Added comment rendering suggestion.

File: internal/handlers/common/projection.go [hunk 3]
```diff
@@ -221,28 +221,270 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
+				doc = d
+			}
+		}
+
+		// when next prefix has an arr use arr,
+		// if it has a document, doc is set by includeProjection.
+		arr, err := includeProjection(path.TrimPrefix(), embeddedSource, doc)
+
+		switch {
+		case errors.Is(err, pathNotExists):
```
Comment (chilagrow) [inline internal/handlers/common/projection.go:297]: `pathNotExists` is still needed here, because we must return without setting document if path does not exist.

**Review by w84thesun — DISMISSED** [review]
LGTM!

File: internal/handlers/common/projection.go [hunk 4]
```diff
@@ -221,28 +221,264 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
-				if inclusion {
-					if docWithoutID.Has(key) {
-						projected.Set(key, must.NotFail(docWithoutID.Get(key)))
-					}
-
-					continue
+			if inclusion {
+				// inclusion projection copies the field on the path from docWithoutID to projected.
+				if _, err = includeProjection(path, docWithoutID, projected); err != nil {
+					return nil, err
```
Comment (rumyantseva) [inline internal/handlers/common/projection.go:231]: How do we process this error? As far as I understand, this error is `pathNotExists`, but I don't see any place where we compare with it or return a specific error code based on it. Also, this row is marked as not covered by tests.

File: internal/handlers/common/projection.go [hunk 5]
```diff
@@ -221,28 +221,264 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
 }
+
+// pathNotExists indicates that no path was found.
+var pathNotExists = errors.New("path not exists")
+
+// includeProjection copies the field on the path from source to projected.
+// When an array is on the path, it returns the array containing any document
+// with the same key. Dot notation with array index path does not include
+// the field unlike document.SetByPath(path).
+// Inclusion projection with non-existent path creates an empty document
```
Comment (rumyantseva) [inline internal/handlers/common/projection.go:254]: From this description, I don't really understand in which cases this function returns errors. Based on this description, a non-existent path doesn't look like an error situation to me.

**Review by rumyantseva — COMMENTED** [review]
Great job clarifying how projection works with dot notation and corner cases! I added a couple of questions about tests and errors.

File: internal/handlers/common/projection.go [hunk 6]
```diff
@@ -221,28 +221,264 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
+			arr.Append(doc)
+		}
+
+		return arr, nil
+	default:
+		return nil, pathNotExists
+	}
+}
+
+// isEmpty returns true if projected document at key is empty. It's empty if:
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/projection.go]: Just great review comment :heart_eyes: , this discovered a bug which I wasn't aware. And we no longer need this function but we need array to handle multiple existing dot notation inclusion. I added `array-three-documents` test data  and test case `DotNotationIncludeTwoArray` to solve the issue 
https://github.com/FerretDB/FerretDB/blob/cf04970ac883fc83c8089eb5ae81bbe6d83eccb3/integration/query_projection_compat_test.go#L157

https://github.com/FerretDB/FerretDB/blob/cf04970ac883fc83c8089eb5ae81bbe6d83eccb3/integration/shareddata/composites.go#L415


Previously, it was checking that if non-empty projection existed, it didn't set the field. So it was only setting it to `v.foo` but not for `v.bar`. This was wrong, as `DotNotationIncludeTwoArray` shows that it has to include both `v.foo` and `v.bar`.

**Review by rumyantseva — APPROVED** [review]

**Review by w84thesun — APPROVED** [review]
LGTM!

**Review by noisersup — APPROVED** [review]
Thanks for the extended comments! LGTM
